### PR TITLE
Add websockets v2

### PR DESCRIPTION
**Pull request by lmoe**
# Description of change
Updates the websocket to use the API models.

Events, commands and validation are now in separate classes, hopefully to make it easier to extend
Publisher has separate event types to support proper model mapping
ISCEvent is now generic throughout the pipeline
wasmlib websocket clients were refactored

## Links to any relevant issues

Be sure to reference any related issues by adding `fixes issue #`.

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Bug fix (a non-breaking change which fixes an issue)
- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation Fix

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [ ] I have followed the [contribution guidelines](https://wiki.iota.org/smart-contracts/contribute) for this project
- [ ] I have performed a self-review of my own code
- [ ] I have selected the `develop` branch as the target branch
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by BugFreeSoftware — APPROVED** [review]

File: packages/publisher/events.go [hunk 1]
```diff
@@ -3,40 +3,65 @@ package publisher
...
+func (e *ISCEvent[T]) String() string {
 	issuerStr := "vm"
-	if e.Issuer != "" {
-		issuerStr = e.Issuer
+	if e.Issuer != nil {
+		issuerStr = e.Issuer.String()
 	}
 	// chainid | issuer (kind):
-	return fmt.Sprintf("%s | %s (%s): %v", e.ChainID, issuerStr, e.Kind, e.Content)
+	return fmt.Sprintf("%s | %s (%s): %v", e.ChainID, issuerStr, e.Kind, e.Payload)
```
Comment (muXxer) [inline packages/publisher/events.go]: do we really need to print the payload here?

File: packages/publisher/events.go [hunk 2]
```diff
@@ -45,49 +70,61 @@ func PublishBlockEvents(blockApplied *publisherBlockApplied, publish func(*ISCEv
...
+	blockEvents, err := blocklog.GetEventsByBlockIndex(blocklogStatePartition, blockIndex, blockInfo.TotalRequests)
 	if err != nil {
-		log.Errorf("unable to get events from a block: %w", err)
+		log.Errorf("unable to get events from a block: %v", err)
 	} else {
-		publish(&ISCEvent{
-			Kind: ISCEventKindSmartContract,
+		triggerEvent(events, events.BlockEvents, &ISCEvent[[]string]{
+			Kind: ISCEventKindBlockEvents,
 			// TODO should be the contract Hname, but right now events are just stored as strings.
```
Comment (muXxer) [inline packages/publisher/events.go:121]: are these TODOs still a thing?

File: plugins/webapi/params.go [hunk 3]
```diff
@@ -14,10 +14,11 @@ type ParametersWebAPI struct {
...
 	Limits struct {
-		Timeout       time.Duration `default:"30s" usage:"the timeout after which a long running operation will be canceled"`
-		ReadTimeout   time.Duration `default:"10s" usage:"the read timeout for the HTTP request body"`
-		WriteTimeout  time.Duration `default:"10s" usage:"the write timeout for the HTTP response body"`
-		MaxBodyLength string        `default:"2M" usage:"the maximum number of characters that the body of an API call may contain"`
+		Timeout                        time.Duration `default:"30s" usage:"the timeout after which a long running operation will be canceled"`
+		ReadTimeout                    time.Duration `default:"10s" usage:"the read timeout for the HTTP request body"`
+		WriteTimeout                   time.Duration `default:"10s" usage:"the write timeout for the HTTP response body"`
+		MaxBodyLength                  string        `default:"2M" usage:"the maximum number of characters that the body of an API call may contain"`
+		MaxTopicSubscriptionsPerClient int           `default:"0" usage:"defines the max amount of subscriptions per client. 0 = deactivated (default)"`
```
Comment (muXxer) [inline plugins/webapi/params.go:21]: Please run the `/scripts/gendoc.sh` script and commit the changes to the default config and the documentation that is generated.

File: plugins/webapi/component.go [hunk 4]
```diff
@@ -171,17 +174,26 @@ func provide(c *dig.Container) error {
...
+			publisher.ISCEventKindReceipt,
+			publisher.ISCEventIssuerVM,
+			publisher.ISCEventKindBlockEvents,
+		}, deps.Publisher, websocket.WithMaxTopicSubscriptionsPerClient(ParamsWebAPI.Limits.MaxTopicSubscriptionsPerClient))
+
 		webapi.Init(
-			Plugin.App().NewLogger("WebAPI/v2"),
+			logger,
 			echoSwagger,
 			hub,
```
Comment (muXxer) [inline plugins/webapi/component.go]: You don't need to pass the hub anymore to the Init.

File: packages/publisher/publisher.go [hunk 5]
```diff
@@ -11,63 +11,64 @@ import (
...
+			Published:      event.New[*ISCEvent[any]](),
 		},
 	}
 
 	return p
 }
 
 // Implements the chain.ChainListener interface.
-// NOTE: Do not block the caller!
+// NOTE: Do not Block the caller!
```
Comment (muXxer) [inline packages/publisher/publisher.go]: find and replace? :D

File: packages/publisher/publisher.go [hunk 6]
```diff
@@ -11,63 +11,64 @@ import (
...
 func (p *Publisher) BlockApplied(chainID isc.ChainID, block state.Block) {
-	p.blockAppliedPipe.In() <- &publisherBlockApplied{chainID: chainID, block: block}
+	p.blockAppliedPipe.In() <- &blockApplied{chainID: chainID, block: block}
 }
 
 // Implements the chain.ChainListener interface.
-// NOTE: Do not block the caller!
+// NOTE: Do not Block the caller!
```
Comment (muXxer) [inline packages/publisher/publisher.go]: find and replace? :D

File: packages/publisher/publisher.go [hunk 7]
```diff
@@ -11,63 +11,64 @@ import (
...
 func (p *Publisher) AccessNodesUpdated(chainID isc.ChainID, accessNodes []*cryptolib.PublicKey) {
 	// We don't need this event.
 }
 
 // Implements the chain.ChainListener interface.
-// NOTE: Do not block the caller!
+// NOTE: Do not Block the caller!
```
Comment (muXxer) [inline packages/publisher/publisher.go]: find and replace? :D

File: packages/webapi/websocket/commands/commandmanager.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+			},
+		},
+		subscriptionManager: subscriptionManager,
+	}
+}
+
+func (c *CommandManager) handleError(err error, client *websockethub.Client, commandType CommandType) error {
+	unwrappedError := errors.Unwrap(err)
+
+	if errors.Is(err, ErrFailedToDeserializeCommand) {
```
Comment (muXxer) [inline packages/webapi/websocket/commands/commandmanager.go]: Use a switch case here :+1:

**Review by muXxer — CHANGES_REQUESTED** [review]

File: packages/publisher/events.go (hunk 2 above)
Comment (lmoe) [reply packages/publisher/events.go:121]: Yeah the events you emit from contracts are just strings. And they not only contain a random string but appended data like the Hname of the contract and so on. We could split the string in the publisher and get the values out of it. But it's pretty hacky and would rather avoid this.

**Pull request merged**